chain/consensus/mir/mine.go

```diff
@@ -1,76 +1,159 @@
-	defer submit.Stop()
+	lastValidatorSetHash, err := m.InitialValidatorSet.Hash()
+	if err != nil {
+		return err
+	}
 
 	for {
+		// Here we use `ctx.Err()` in the beginning of the `for` loop instead of using it in the `select` statement,
+		// because if `ctx` has been closed then `api.ChainHead(ctx)` returns an error,
+		// and we will be in the infinite loop due to `continue`.
+		if ctx.Err() != nil {
+			log.Debug("Mir miner: context closed")
+			return nil
+		}
 		base, err := api.ChainHead(ctx)
 		if err != nil {
 			log.Errorw("failed to get chain head", "error", err)
 			continue
 		}
 
-		// Miner (leader) for an epoch is assigned deterministically using round-robin.
-		// All other validators use the same Miner in the block.
-		epochMiner := sm.Validators[int(base.Height())%len(sm.Validators)].Addr
 		nextEpoch := base.Height() + 1
 
 		select {
-		case <-ctx.Done():
-			log.With("epoch", nextEpoch).Debug("Mir miner: context closed")
-			return nil
+
 		case err := <-mirErrors:
-			return fmt.Errorf("miner consensus error: %w", err)
-		case batch := <-mirHead:
-			msgs, crossMsgs := sm.GetMessages(batch)
+			// return fmt.Errorf("miner consensus error: %w", err)
+			//
+			// TODO: This is a temporary solution while we are discussing that issue
+			// https://filecoinproject.slack.com/archives/C03C77HN3AS/p1660330971306019
+			panic(fmt.Errorf("miner consensus error: %w", err))
+
+		case newMembership := <-m.App.MembershipNotify:
+			if err := m.ReconfigureMirNode(ctx, newMembership); err != nil {
+				log.With("epoch", nextEpoch).Errorw("reconfiguring Mir failed", "error", err)
+				continue
+			}
+
+		case <-reconfigure.C:
+			// Reconfiguration is not used in the rootnet.
+			if m.SubnetID == address.RootSubnet {
+				continue
+			}
+			//
+			// Send a reconfiguration transaction if the validator set in the actor has been changed.
+			//
+
+			// NOTE: You must unset the environment variable in tests if you use Mir in the rootnet and in a subnet.
+			// Should we support passing validators via the environment variable?
+			// If yes then we should Implement a sophisticated way to separate getting validator
+			// set via environment variable and subnet actor.
+			// A membership is passed to Mir via the environment variable for the rootnet (for demo and debugging purposes)
+			// and via the subnet actor for a subnet. The environment variable is read first.
+			// We have tests where Mir runs in the rootnet and a subnet simultaneously.
+			// So if you don't unset the variable after instantiation Mir in the rootnet
+			// the subnet Mir miner cannot get membership.
+			// The environment variable must be empty because otherwise it will be prioritized for a subnet.
+			newValidatorSet, err := getSubnetValidators(ctx, m.SubnetID, api)
+			if err != nil {
+				log.With("epoch", nextEpoch).Warnf("failed to get subnet validators: %v", err)
+				continue
+			}
+
+			newValidatorSetHash, err := newValidatorSet.Hash()
```

Comment:
hahaha this is an interesting way of checking if there's a diff. Do you think that marshaling and computing the hash is more performant than looping through the map to check the diff'ed elements?

Reply:
Implemented your suggestion - https://github.com/filecoin-project/eudico/pull/234/commits/12fceca092df2fbd3caedb2714c3aff4c64f11da

Reply:
We don't use maps. ValidatorSet is a slice.

> How big we thing this maps can be?

it depends on subnets.

Reply:
Iterate the lists (sorry, typo).

Reply:
Yes, now it works by iterating the lists.

Reply:
This is such a rare event (once a timer ticker) that I'd be extremely surprised if it had any impact whatsoever on the performance, regardless of the method used. I wouldn't consider performance to be an issue here (unless extremely wasteful) and prioritize the easiest-to-understand solution (both hashing and iterating seem fine to me, I guess it's a matter of personal taste).